apps/prairielearn/src/pages/studentInstanceQuestionExam/studentInstanceQuestionExam.js

```diff
@@ -93,87 +95,140 @@ function processSubmission(req, res, callback) {
+      }
+      if (res.locals.assessment.group_work) {
+        const groupConfig = await groupAssessmentHelper.getGroupConfig(res.locals.assessment.id);
+        if (groupConfig.has_roles) {
+          const result = await groupAssessmentHelper.getQuestionPermissions(
+            res.locals.assessment_question.id,
+            res.locals.user.user_id,
+          );
+          // Users without the correct roles cannot submit
+          if (!result.can_submit) {
+            return next(
+              error.make(
+                403,
+                'Current group roles have no permission to submit this question',
+                res.locals,
+              ),
+            );
+          }
+
+          // Get group information
+          const groupId = await groupAssessmentHelper.getGroupId(
+            res.locals.assessment.id,
+            res.locals.user.user_id,
+          );
+          const groupInfo = await groupAssessmentHelper.getGroupInfo(groupId, groupConfig);
+          const { validationErrors, rolesAreBalanced, usersWithoutRoles } = groupInfo.rolesInfo;
+
+          // Prevent submissions if role configuration is invalid
+          if (validationErrors.length > 0 || !rolesAreBalanced || usersWithoutRoles.length > 0) {
+            return next(
+              error.make(
+                403,
+                'Current group role configuration is invalid. Submissions are disabled',
+                res.locals,
+              ),
+            );
+          }
+        }
+      }
+
+      await util.promisify(processSubmission)(req, res);
+      res.redirect(req.originalUrl);
+    } else if (req.body.__action === 'timeLimitFinish') {
+      // TODO: Does this need to be turned async?
+
+      // Only close if the timer expired due to time limit, not for access end
+      if (!res.locals.assessment_instance_time_limit_expired) {
+        return res.redirect(req.originalUrl);
+      }
+
+      const requireOpen = true;
+      const closeExam = true;
+      const overrideGradeRate = false;
+      assessment.gradeAssessmentInstance(
+        res.locals.assessment_instance.id,
+        res.locals.authn_user.user_id,
+        requireOpen,
+        closeExam,
+        overrideGradeRate,
+        function (err) {
+          if (ERR(err, next)) return;
+          res.redirect(
+            res.locals.urlPrefix +
+              '/assessment_instance/' +
+              res.locals.assessment_instance.id +
+              '?timeLimitExpired=true',
+          );
+        },
+      );
+    } else if (req.body.__action === 'attach_file') {
+      util.callbackify(studentInstanceQuestion.processFileUpload)(
+        req,
+        res,
+        function (err, variant_id) {
+          if (ERR(err, next)) return;
+          res.redirect(
+            res.locals.urlPrefix +
+              '/instance_question/' +
+              res.locals.instance_question.id +
+              '/?variant_id=' +
+              variant_id,
+          );
+        },
+      );
```

Comment:
Now that this is an async function, you can avoid the use of `util.callbackify` here and just directly await the result of `studentInstanceQuestion.processFileUpload(...)`.